Prevents node from starting given persisted network state is inconsistent with configuration

abailly-iohk opened the pull request:
To fix #1174  we choose to `die` when the loaded persisted state for the network acknowledgments is not consistent with the configured parties, eg. when they don't have the same size. This is a bit lame as we don't really check much and it could very well be the case the user restarts a node with a different configuration (eg. different parties) but same number and got tripped by this issue. 

![deeper](https://github.com/input-output-hk/hydra/assets/79840582/7ff1ac94-2da7-43b2-b538-96038d083081)

---

<!-- Consider each and tick it off one way or the other -->
* [x] CHANGELOG updated or not needed
* [x] Documentation updated or not needed
* [x] Haddocks updated or not needed
* [x] No new TODOs introduced or explained herafter

ffakenz committed:
Draft e2e spec

abailly committed:
Capture stderr from child process

This is needed in order to see in our test the node has died with a
misconfiguration, which is not the problem we are trying to solve but
which is an important information to have.

ffakenz committed:
Fix draft e2e spec so it fails for the right reason

Also renaming the spec to better match the scenarion.

ffakenz committed:
Check when acks dir missmatches with the peers configuration

ffakenz committed:
Extend spec to check we can resume a well configured peer

abailly committed:
Parameterize peers connectivity check

This is not strictly needed but allows for fine tuning the time we
wait for connection depending on the exact context of the check, and
not blindly on the number of nodes.

abailly committed:
Provide a more specific exception for network persistence mismatch

ffakenz committed:
Trace params mismatch missconfiguration error when SavedNetworkPartiesInconsistent occur

ffakenz committed:
Update logs schema with new params mismatch

ffakenz commented on hydra-node/src/Hydra/Logging.hs:
```diff
@@ -140,7 +140,7 @@ traceInTVar ::
   Tracer m msg
 traceInTVar tvar = Tracer $ \msg -> do
-  envelope <- mkEnvelope "" msg
+  envelope <- mkEnvelope "TEST" msg
```
@abailly-iohk do we want to keep this?

abailly committed:
Encapsulate file name in a more explicit function

ffakenz committed:
Add spec namespace when showing logs on failure

ffakenz committed:
Update CHANGELOG

ffakenz committed:
Minor formatting fixes

ch1bo commented on hydra-node/src/Hydra/Node/Network.hs:
```diff
@@ -131,6 +136,30 @@ withNetwork tracer connectionMessages configuration callback action = do
 
+-- | Create `MessagePersistence` handle to be used by `Reliability` network layer.
+--
+-- This function will `throw` a `ConfigurationMismatch` exception if:
```
Wrong haddock, I think this should read `ParameterMismatch`

ch1bo commented on hydra-node/src/Hydra/Logging.hs:
```diff
@@ -128,19 +128,21 @@ withTracerOutputTo hdl namespace action = do
-traceInTVar tvar = Tracer $ \msg -> do
-  envelope <- mkEnvelope "" msg
+traceInTVar tvar namespace = Tracer $ \msg -> do
+  envelope <- mkEnvelope namespace msg
```
What is the benefit of having a namespace; where did we need it? I see it required a lot of code changes and makes the test boilerplate bigger.

ffakenz replied:
while working on the e2e test, we encountered a situation where resuming a Hydra node (after reconfiguring a party) became confusing. This confusion arose because the logs were overwritten by the second launch of the instance.
wdyt?

ffakenz replied:
imo this tidy result from a need originated by this PR.

I can do it if you think differently, as it is self contained in a single commit so no big deal.

let me know your thoughts.

ch1bo commented on hydra-node/test/Hydra/NetworkSpec.hs:
```diff
@@ -52,8 +54,14 @@ spec = do
     prop "can roundtrip CBOR encoding/decoding of Hydra Message" $ prop_canRoundtripCBOREncoding @(Message SimpleTx)
     roundtripAndGoldenSpecs (Proxy @(Message SimpleTx))
 
+  describe "Message Persistence" $ do
```
The test case name indicates ("it throws..") that this is a unit test about a specific function. I would suggest to use exactly that function name in the `describe` scope. This helps in finding relevant tests and eases future renaming of this function.

```suggestion
  describe "configureMessagePersistence" $ do
```

ch1bo commented on hydra-node/test/Hydra/NetworkSpec.hs:
```diff
@@ -52,8 +54,14 @@ spec = do
+    it "throws ParameterMismatch when configuring given number of acks does not match number of parties" $ do
+      withTempDir "persistence" $ \dir -> do
+        writeFile (acksFile dir) "[0,0,0]"
+        configureMessagePersistence @_ @Int nullTracer dir 4 `shouldThrow` (const True :: Selector ParameterMismatch)
```
Should: avoid writing the internal representation here.

Instead, we can express ourselves by using the interfaces provided of the "system under test", here: `configureMessagePersistence`

```suggestion
        MesagePersistence{saveAcks} <- configureMessagePersistence @_ @Int nullTracer dir 3
        saveAcks (fromList [0, 0, 0])
        configureMessagePersistence @_ @Int nullTracer dir 4 `shouldThrow` (const True :: Selector ParameterMismatch)
```

ch1bo commented on hydra-node/json-schemas/logs.yaml:
```diff
@@ -720,6 +720,22 @@ definitions:
               Parties configured as the node argument.
+      - title: SavedNetworkPartiesInconsistent
+        description: >-
+          The configured peer list does not match with the value from the loaded /acks state.
```
It is a bit surprising to see the concrete path `/acks` here. IMO it has a high chance to bitrot here.

Should: not be so specific here

---

If the intention was to explain what to do in case you see this error in the logs, I think this would have not been discovered anyways by anyone (logs.yaml is not really part of our documentation).

ch1bo reviewed (changes requested):
Surprisingly large change. I think we can (and should) reduce the number of mentions of `acks` and other internal details of the network layer.

One thing I am missing is any concrete instructions to the user. How would they know that they need to delete the `<persistence-dir>/acks` file? To solve this, I would still `die` with an exception, but implement https://hackage.haskell.org/package/base-4.19.0.0/docs/Control-Exception.html#v:displayException of the thrown exception accordingly to give a human-friendly english instruction to delete `acksFile`

ffakenz committed:
Use the function name in the describe scope when unit testing about a specific function

This helps in finding relevant tests and eases future renaming of this function.

Co-authored-by: Sebastian Nagel <ch1bo@users.noreply.github.com>

ghost commented:
You're right there's some instructions missing and we should provide better error messages to the user should we crash with an exception. However:
1. this is not specific to this PR, there are [other places](https://github.com/input-output-hk/hydra/blob/00e9743f3e05243fc57991dc42540c986b0620bb/hydra-node/exe/hydra-node/Main.hs#L87) where we throw  exceptions without caring about the user-facing message
2. I would not use `displayException` for this but would rather reuse our [`explain`](https://github.com/input-output-hk/hydra/blob/00e9743f3e05243fc57991dc42540c986b0620bb/hydra-node/src/Hydra/Options.hs#L639) function: It avoids "scattering" error messages in various places where the exceptions are defined, and is much more amenable to localisation and customisation, eg. should we want to display this error using JSON, or even i18n. I think it's the responsibility of the user interface to provide sensible feedback to the user.

Doing so here would increase even more the surface of this PR so I would like to propose we address this particular issue in a later PR.

ffakenz committed:
Improve network haddock on configureMessagePersistence

ffakenz committed:
Fix logs description for SavedNetworkPartiesInconsistent by not being too concrete

ffakenz committed:
Improve network spec by avoid writing the internal representation when saving acks

Instead, we can use the interface provided for 'the system under test', here: configureMessagePersistence.

ch1bo commented:
> You're right there's some instructions missing and we should provide better error messages to the user should we crash with an exception. However:
> 
>     1. this is not specific to this PR, there are [other places](https://github.com/input-output-hk/hydra/blob/00e9743f3e05243fc57991dc42540c986b0620bb/hydra-node/exe/hydra-node/Main.hs#L87) where we throw  exceptions without caring about the user-facing message
> 
>     2. I would not use `displayException` for this but would rather reuse our [`explain`](https://github.com/input-output-hk/hydra/blob/00e9743f3e05243fc57991dc42540c986b0620bb/hydra-node/src/Hydra/Options.hs#L639) function: It avoids "scattering" error messages in various places where the exceptions are defined, and is much more amenable to localisation and customisation, eg. should we want to display this error using JSON, or even i18n. I think it's the responsibility of the user interface to provide sensible feedback to the user.
> 
> 
> Doing so here would increase even more the surface of this PR so I would like to propose we address this particular issue in a later PR.

I'm okay in not doing that in this PR.. it's just something which I thought would be nice small step in less confusing error messages.

I see your point with different encodings and internationalization (although YAGNI?), but would want to avoid arbitrary centralization of otherwise unreleated exceptions. The case we have here is a good example, I would want to retain information that some file at path "<persistence-dir>/acks" should be deleted to the component which decides that we do persist into a file at that path. Related to this is the IMO unnecessary addition of `SavedNetworkPartiesInconsistent` to `ParameterMismatch`, which has a completely different origin.

That being said, I can approve and might just PR what I think should be myself :)

ghost commented:
> I see your point with different encodings and internationalization (although YAGNI?), but would want to avoid arbitrary centralization of otherwise unreleated exceptions. 

Agreed, this is YAGNI and I just raised this point to highlight the current need to avoid putting arbitrary strings into places where they do not belong: Defining and using the (structured) error is one thing, presenting this issue in an understandable way is a different concern.

> The case we have here is a good example, I would want to retain information that some file at path "/acks" should be deleted to the component which decides that we do persist into a file at that path. 

I agree, this is currently missing from the exception and should be added.

> Related to this is the IMO unnecessary addition of SavedNetworkPartiesInconsistent to ParameterMismatch, which has a completely different origin.

That's where our views differ: This issue comes from a discrepancy between the persisted state and the configuration passed to the node which percolates down to the network layer, hence I think it's legitimate to make it part of those exceptions related to configuration "errors". We could refine the hierarchy more perhaps, but at this stage it's definitely YAGNI :)

To be clear, what I have in mind is something like that in the `Main` module:

```
main :: IO ()
main = do
   parseOptions >>= try . configureNode >>= \case
      Left (e :: PreFlightCheckError) -> die (explain e)
      Right node -> runNode node
```
